### PR TITLE
build: upgrade golangci-lint and run lint in Github Action

### PR DESCRIPTION
**Pull request by lodrem**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

1. Upgrade `golangci-lint` to `v1.50.1`
2. Run `lint` in Github Action: better performance and no hack for different environments.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -400,3 +400,12 @@ endif
...
 	hack/deploy-cluster-capz.sh
+
+##@ Tools
+
+LINTER = $(shell pwd)/bin/golangci-lint
+LINTER_VERSION = v1.50.1
+.PHONY: golangci-lint
+golangci-lint:  ## Download golangci-lint locally if necessary.
+	@echo "Installing golangci-lint"
+	@test -s $(LINTER) || curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(shell pwd)/bin $(LINTER_VERSION)
```
Comment (lzhecheng) [inline Makefile:411]:
```suggestion
	@test -s $(LINTER) || curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/$(LINTER_VERSION)/install.sh | sh -s -- -b $(shell pwd)/bin $(LINTER_VERSION)
```
For install.sh script, I think it is better to pin to a specific version like `v1.50.1`
https://raw.githubusercontent.com/golangci/golangci-lint/v1.50.1/install.sh

File: Makefile [hunk 2]
```diff
@@ -317,11 +317,11 @@ ifdef JUNIT
 endif
 
 .PHONY: test-check
-test-check: test-lint test-boilerplate test-helm verify-vendor-licenses ## Run all static checks.
+test-check: test-boilerplate test-helm verify-vendor-licenses ## Run all static checks.
 
-.PHONY: test-lint
-test-lint: ## Run golint test.
-	hack/verify-golint.sh
+.PHONY: lint
```
Comment (lzhecheng) [inline Makefile:322]: Nit: I prefer the old name `test-lint`. No strong opinion.

File: .github/workflows/lint.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+name: Code Lint
+
+on:
+  workflow_dispatch:
+  push:
+  pull_request:
+
+jobs:
+  lint:
```
Comment (lzhecheng) [inline .github/workflows/lint.yaml:9]: Maybe better to capitalize it for consistency.

**Review by lzhecheng — COMMENTED** [review]
Overall lgtm. A few nits.

File: Makefile (hunk 2 above)
Comment (lodrem) [reply Makefile:322]: To me, `test` doesn't link to `lint` and I want to keep it short. 😄

**Comment by elmiko** [comment]
fwiw, it would have been nice to mention the makefile target change in the description, perhaps in the "Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:" section. this change broke our tooling. not a huge deal, but just sharing.

**Comment by lzhecheng** [comment]
> fwiw, it would have been nice to mention the makefile target change in the description, perhaps in the "Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:" section. this change broke our tooling. not a huge deal, but just sharing.

Thank you for sharing this. Yes, it is worth mentioning in the description.

**Pull request merged**